[Step1]엔티티 매핑 PR 요청 드립니다.

re-mora opened the pull request:
안녕하세요 :) 교육생 최용석입니다.

1단계 엔티티 매핑 미션 구현이 완료되어, 코드 리뷰 요청드립니다.

미흡한 부분이 있다면 피드백 부탁드립니다!

re-mora committed:
docs: 요구사항 분석 및 구현 기능 목록 작성

re-mora committed:
feat(configurations): application 구동에 필요한 설정 추가

- 원활한 TC 수행을 위한 auto ddl 설정 적용 : create-drop
- 실행 쿼리 출력을 위한 설정 적용

re-mora committed:
feat: 회원 테이블을 중심으로 회원 Entity 매핑 및 Repository TC 작성

User 테이블과 User Entity 매핑
- @UniqueConstraint를 이용한 유니크 제약조건 설정
- @Id를 이용한 PK 컬럼 매핑 및 PK 컬럼 값 설정 전략 적용
- @GeneratedValue IDENTITY 설정을 이용한 PK 컬럼 값 생성 전략 적용
  - Insert 시점에 DB에서 값을 부여
- @Column의 속성을 이용하여 컬럼의 세부 속성 매핑

TC 작성
- UserEntity 및 UserRepository 동작 확인을 위한 TC 작성
- TC 작성 시, 반복되는 메서드 추출

re-mora committed:
feat: 질문 테이블을 중심으로 질문 Entity 매핑 및 Repository TC 작성

Question 테이블과 Question Entity 매핑
- @Id를 이용한 PK 컬럼 매핑 및 PK 컬럼 값 설정 전략 적용
- @GeneratedValue IDENTITY 설정을 이용한 PK 컬럼 값 생성 전략 적용
  - Insert 시점에 DB에서 값을 부여
- @Column의 속성을 이용하여 컬럼의 세부 속성 매핑
- @Lob을 이용하여 clob으로 정의된 컬럼 매핑

TC 작성
- Entity 및 Repository 동작 확인을 위한 TC 작성
- TC 작성 시, 반복되는 메서드 추출

re-mora committed:
feat: 답변 테이블을 중심으로 답변 Entity 매핑 및 Repository TC 작성

Answer 테이블과 Answer Entity 매핑
- @Id를 이용한 PK 컬럼 매핑 및 PK 컬럼 값 설정 전략 적용
- @GeneratedValue IDENTITY 설정을 이용한 PK 컬럼 값 생성 전략 적용
  - Insert 시점에 DB에서 값을 부여
- @Column의 속성을 이용하여 컬럼의 세부 속성 매핑
- @Lob을 이용하여 clob으로 정의된 컬럼 매핑

TC 작성
- Entity 및 Repository 동작 확인을 위한 TC 작성
- TC 작성 시, 반복되는 메서드 추출

re-mora committed:
feat: 삭제이력 테이블을 중심으로 삭제이력 Entity 매핑 및 Repository TC 작성

delete_history 테이블과 DeleteHistory Entity 매핑
- @Id를 이용한 PK 컬럼 매핑 및 PK 컬럼 값 설정 전략 적용
- @GeneratedValue IDENTITY 설정을 이용한 PK 컬럼 값 생성 전략 적용
  - Insert 시점에 DB에서 값을 부여
- @Enumerated를 이용한 열거형 타입 매핑

TC 작성
- Entity 및 Repository 동작 확인을 위한 TC 작성

seokhongkim commented on docs/step1-README.md:
```diff
@@ -0,0 +1,173 @@
+
+### 구현 기능 목록 : TODO List
+- Entity 매핑
+- [x] 회원 Entity 설계
```
문서를 꼼꼼하게 작성해주셨네요! 👍 
문서를 잘 작성해주시는 것은 좋은 습관이라고 생각합니다.
추후 유지보수까지 고민해본다면, 위의 DDL 과 중복되는 부분, 테스트 코드의 내용과 중복되는 부분에 대해서 관리 방안도 같이 고민해보시면 더 좋은 문서 작성에 도움이 될 것 같아요!

seokhongkim commented on docs/step1-README.md:
```diff
@@ -0,0 +1,173 @@
+  - DB 중심으로 객체를 설계 했을때 발생하는 문제점
+    - 객체간 참조가 어려운 구조, 외래키에 해당하는 연관 객체를 다시 조회
+    
+### JAP 엔티티 설계
```
오타인 것 같아요!

re-mora replied:
피드백 적용 커밋에 포함하도록 하겠습니다!

seokhongkim commented on docs/step1-README.md:
```diff
@@ -0,0 +1,173 @@
+   - Enum Type의 저장 방식이 순서 저장 방식인 Orginal 설정이 아닌 값 저장 방식인 String 적용 여부 확인
+
+---
+### 학습 키워드
```
학습 키워드를 정리하시는 것도 좋네요! 👍

seokhongkim commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -1,17 +1,40 @@
     private boolean deleted = false;
 
+    @Column(nullable = false)
+    private LocalDateTime createdAt = LocalDateTime.now();
```
생성일시와 수정일시는 `Answer`, `Question`, `User` 에서 모두 필요할 것 같아요.
학습을 위해 `MappedSuperclass` 를 찾아보시는 건 어떨까요?
`MappedSuperclass` 를 학습하셨는데, 필요성을 느끼지 못하시는 거라면 말씀해주세요!

re-mora replied:
주어진 Step1의 미션 틀에서 최대한 구조 변경 없이 진행하려는 생각에 Auditor는 고려하지 않게 된것 같습니다.
말씀해주신 것 처럼 JPA Auditing을 이용하여 Entity의 공통 속성 추출 및 매핑 정보를 상속하는 부분을 적용 후, 피드백 반영 커밋에 포함하도록 하겠습니다!

seokhongkim commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -1,17 +1,40 @@
     private boolean deleted = false;
 
+    @Column(nullable = false)
+    private LocalDateTime createdAt = LocalDateTime.now();
```
데이터 생성, 저장 일시를 자동으로 관리할 수 있는 `Jpa Audit` 기능을 학습해보시면 좋을 것 같습니다. 😀

seokhongkim commented on src/main/java/qna/domain/DeleteHistory.java:
```diff
@@ -2,30 +2,58 @@
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
     private Long id;
+
+    @Enumerated(EnumType.STRING)
```
`Enumerated` 에 `String` 으로 지정해주신 것 좋습니다. 👍

seokhongkim commented on src/main/java/qna/domain/DeleteHistory.java:
```diff
@@ -2,30 +2,58 @@
+    }
+
     @Override
     public boolean equals(Object o) {
```
`DeleteHistory` 에만 `equals` 가 오버라이드 되어 있네요.
JPA Entity 의 equals, hashcode 는 어떻게 구현하는 게 좋을까요? 여러 방법이 있으니 `JPA Entity equals` 와 같은 키워드로 학습해보시면 좋을 것 같습니다!

re-mora replied:
이 부분에 대해서 평소에 기계적으로 구현하고 지나쳤던 부분인데, 꼼꼼하게 살펴봐주신 덕분에 다시 한번 살펴보는 계기가 되었습니다. 👍 
각 Entity의 Equals And Hashcode를 구현하고 피드백 반영 커밋에 포함하도록 하겠습니다.

seokhongkim commented on src/main/java/qna/domain/User.java:
```diff
@@ -1,18 +1,47 @@
 
+@Entity
+@Table(name = "user",
+    uniqueConstraints = @UniqueConstraint(
```
unique constraint 설정을 명시적으로 해주신 것 좋습니다. 👍

seokhongkim commented on src/test/java/qna/domain/AnswerRepositoryTest.java:
```diff
@@ -0,0 +1,200 @@
+import qna.generator.UserGenerator;
+
+/**
+ * @author : choi-ys
```
test javadoc 은 일부러 추가하신걸까요? 커밋 로그로 이력 관리가 될 것 같은데, javadoc 을 따로 추가하신 이유가 있을까요?

re-mora replied:
이 부분도 IDE에 기계적으로 설정하여 의식하지 못한 부분인 것 같습니다.
말씀해주신 것처럼 해당 주석이 표현하는 정보를 git을 통해 확인할 수 있음을 인지하고 관련 코드와 설정은 제거하였습니다!

seokhongkim commented on src/test/java/qna/domain/AnswerTest.java:
```diff
@@ -1,6 +1,64 @@
+
+        // Then
+        assertAll(
+            () -> assertThat(given.getId()).as("IDENTITY 전략에 의해 DB에서 생성되는 PK값의 Null 여부").isNull(),
```
이 테스트에서는 Entity 의 키 생성 전략에 따른 값을 검증하지는 않는 것 같습니다.
`Answer` 생성자에서 `id` 를 특정 값으로 지정하더라도, Entity 의 키 생성 전략은 `Identity` 를 사용할 수 있습니다.
이 테스트의 목적은 생성자로 생성된 객체가 있을 때 기본값이 있는 필드와 없는 필드를 검증하는 것일까요?

re-mora replied:
해당 테스트 작성 시, 생성자를 통해 주입한 값이 할당 되었는지, 기본값이 있는 필드에 기본값이 적용 되었는지를 검증하려 하였습니다.
어떤 것들을 검증해야 하는지가 명확하지 않다보니, 기계적으로 객체 생성에 대한 테스트 코드를 작성한 것 같습니다.

리뷰해 주신 내용을 확인 하고 해당 테스트 코드를 보며 어떤 것을 검증하고 싶은지, 객체 생성에 대한 테스트 작성이 의미가 있는지 고민을 많이 하게 되었습니다. 

검증 항목을 명확하게 하는 연습에 많은 필요성을 느끼는 리뷰였습니다. 꼼꼼하게 살펴봐주셔서 감사합니다 :)

seokhongkim commented on src/test/java/qna/domain/AnswerTest.java:
```diff
@@ -1,6 +1,64 @@
+        given.toQuestion(question);
+
+        // Then
+        assertThat(given.getQuestionId()).isEqualTo(question.getId());
```
이때 `questionId` 에는 어떤 값이 들어있을까요?

re-mora replied:
questionId 항목은 null값을 가지게 될 것 같습니다.
해당 테스트 코드 역시 어떤 것을 검증하고 싶은지에 대한 고민이 많이 부족했던 것 같습니다.

seokhongkim commented on src/test/java/qna/domain/DeleteHistoryRepositoryTest.java:
```diff
@@ -0,0 +1,84 @@
+
+    @Test
+    @DisplayName("질문 삭제 이력 저장")
+    public void saveQuestionDeleteHistory() {
```
두개의 테스트가 굉장히 비슷한 것 같습니다.
`DeleteHistory` 를 저장할 때 아래 두 내용을 검증하고 싶으신 것이 맞을까요?
1. `id` 가 잘 들어있는지
2. `Question`, `Answer` 타입으로 `DeleteHistory` 가 잘 저장되는지

re-mora replied:
테스트 케이스의 검증항목이 명확하게 작성되지 않은 것 같습니다.
2번에서 말씀해주신 것처럼 어떤 타입의 삭제 이력인지 검증 하는 부분을 피드백 적용 커밋에 포함하도록 하겠습니다.

seokhongkim reviewed (changes requested):
용석님 안녕하세요.
JPA 리뷰어 김석홍입니다. 만나서 반갑습니다. 😀 
첫 리뷰부터 늦어졌는데 양해해주셔서 감사합니다.

README 문서를 꼼꼼하게 작성하기 위해 노력하신 것 같고, 커밋을 봤을 때 엔티티 단위로 나눠서 작업하기 위해 노력하신 것 같습니다. 👍 

리뷰 코멘트를 조금 남겼으니 확인 부탁드립니다. 🙏

re-mora committed:
refactor: JPA Auditing을 이용한 Entity의 공통 속성 추출 및 매핑 정보 상속

코드 변경에 따른 TC 검증 부 수정

Docs 내 잘못 표현된 오타 수정

re-mora committed:
chore(polish): git을 통해 이미 관리 되고 있는 이력용 javadoc 주석 제거

Java Class 생성 시 주석 템플릿 주입 관련 IntelliJ 설정 제거

re-mora committed:
refactor: Entity 객체 수정 - 코드리뷰 피드백 적용

Entity 객체에 Equals and Hashcode 적용
- 테이블의 각 row를 표현하는 Entity의 경우, DB에 의해 유일성이 보장되므로 비지니스 키(userId)와 기본 키를 이용한 equals and hashcode 재 정의

Entity 객체에 프록시 사용을 고려한 최소 접근 제한지인 Protected 기본 생성자 추가

DeleteHistoryRepositoryTest에 작성된 검증 항목이 명확하지 않은 객체 생성 TC의 검증 부 수정

seokhongkim commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -80,14 +100,31 @@ public void setDeleted(boolean deleted) {
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(id, writerId, questionId, contents, deleted);
```
`equals` 와 `hashCode` 에서 사용하는 필드가 다르네요.
`id` 는 같으나 다른 필드에 다른 값이 있는 경우, `equals` 와 `hashCode` 비교가 다른 결과를 발생시킬 수 있겠네요!

seokhongkim commented on src/main/java/qna/domain/BaseEntity.java:
```diff
@@ -0,0 +1,29 @@
+@MappedSuperclass
+public class BaseEntity {
+
+    @CreatedDate
```
Audit 기능을 사용했을 때 `CreatedDate` 와 `LastModifiedDate` 가 어떤 식으로 값이 설정되는지도 학습해보시면 좋을 것 같습니다.
하나의 트랜잭션 내에서 다른 엔티티 여러개를 생성하거나 수정할 때 각 엔티티의 생성일시, 수정일시가 어떻게 저장되는지도 보시면 좋을 것 같아요.
이 부분을 학습, 테스트 해보시면 언제 Audit 기능을 사용하고, 사용하지 않는 게 좋을지 결정하시는 데 도움이 되리라 생각합니다.

seokhongkim reviewed (approved):
용석님, JPA 1단계 미션 잘 마무리해주셨습니다. 👍 

간단한 코멘트를 남겼는데, 이 부분은 2단계 미션 진행하실 때 같이 고민하고 보셔도 좋을 것 같습니다.
1단계 PR은 이만 머지하도록 하겠습니다.

2단계 미션에서 뵙겠습니다. 🙏

re-mora commented:
> 용석님, JPA 1단계 미션 잘 마무리해주셨습니다. 👍
> 
> 간단한 코멘트를 남겼는데, 이 부분은 2단계 미션 진행하실 때 같이 고민하고 보셔도 좋을 것 같습니다. 1단계 PR은 이만 머지하도록 하겠습니다.
> 
> 2단계 미션에서 뵙겠습니다. 🙏

늦은 시간 까지 확인해 주셔서 감사합니다. 추가 코멘트 해주신 부분 확인해서 2단계 미션에서 뵙겠습니다 :)
감사합니다 👍
